### PR TITLE
Optimise multiply chains

### PR DESCRIPTION
**Pull request by ltratt**
The main part of this commit is https://github.com/ykjit/yk/commit/0145e7ba45913a694052c9d4731b764a2e8be5fa, which allows us to optimise chains of multiplications of constants, but first of all I had to fix a "what was I thinking?" bug (https://github.com/ykjit/yk/commit/4209c61cb914eb83aaadc70dee9011328d2a1192).

### CODE REVIEW

File: ykrt/src/compile/jitc_yk/opt/simple.rs [hunk 1]
```diff
@@ -45,7 +45,7 @@ fn opt_mul(
                 } else if y == 1 {
                     // Replace `x * 1` with `x`.
                     m.replace(inst_i, Inst::ProxyInst(mul_inst));
-                } else if y % 2 == 0 {
+                } else if y & (y - 1) == 0 {
```
Comment (ptersilie) [inline ykrt/src/compile/jitc_yk/opt/simple.rs]: Clever!

File: ykrt/src/compile/jitc_yk/opt/simple.rs [hunk 2]
```diff
@@ -34,24 +34,50 @@ fn opt_mul(
...
-        | (Operand::Const(mul_const), Operand::Local(mul_inst)) => {
+        (Operand::Local(mut mul_inst), Operand::Const(mul_const))
+        | (Operand::Const(mul_const), Operand::Local(mut mul_inst)) => {
             let old_const = m.const_(mul_const);
-            if let Some(y) = old_const.int_to_u64() {
-                if y == 0 {
+            if let Some(old_val) = old_const.int_to_u64() {
+                let mut new_val = old_val;
+                // If we've got `%2: mul %1, xi8` then see if `%1` is of the form `mul %0, yi8`: if so
+                // we've got a chain that's `%2: %0 x*y`. We can thus "skip" the intermediate `mul`
```
Comment (ptersilie) [inline ykrt/src/compile/jitc_yk/opt/simple.rs:43]: I think the chain should be `%2: %0*y*x`.

**Comment by ptersilie** [comment]
Just a typo. I see that due to the structure of traces, this will naturally merge multiple (>2) multiplications despite the optimisation only looking at 2 at a time. However, I believe a mul without constants in the middle can block some merging, e.g.
```
%1 = %0 * 2;
%2 = %1 * %1;
%3 = %2 * 2;
```
which otherwise could be optimised to:
```
%2 = %1 * %1;
%3 = %2 * 4;
```

File: ykrt/src/compile/jitc_yk/opt/simple.rs (hunk 2 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/opt/simple.rs:43]: `*` is commutative, so the two are equivalent.

**Comment by ltratt** [comment]
> However, I believe a mul without constants in the middle can block some merging, e.g.

Correct. I suspect we'll want to do a more clever tracking mechanism to spot such things, but we can get there bit-by-bit.

File: ykrt/src/compile/jitc_yk/opt/simple.rs (hunk 2 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/opt/simple.rs:43]: Fine, but you're still missing a `*`.

Comment (ltratt) [reply ykrt/src/compile/jitc_yk/opt/simple.rs]: Fixed in d668e29917b0dfc697329338a83520193b1e82fa.

**Comment by ptersilie** [comment]
Please squash.

**Comment by ltratt** [comment]
Squashed.

**Pull request merged**